webhooks/sentry: Rewrite the sentry webhook for the latest SDKs.

Hypro999 opened the pull request:
This commit and PR are a WIP and should NOT be merged into master yet. I still need to test some of the issue events manually (well, first I need to get the fixtures for them). Then I need to write the automated tests and update the docs.

I also want to try out a few more scenarios and the SDK payloads in some other languages first to make sure that I'm not making any assumptions that don't universally hold true.

Here's what the messages look like so far:
![exception_event](https://user-images.githubusercontent.com/29123352/80629314-4bbeff00-8a70-11ea-8853-d126a815a998.png)
![message_event](https://user-images.githubusercontent.com/29123352/80629321-4d88c280-8a70-11ea-93f4-cbc7b0850bc4.png)
![issue](https://user-images.githubusercontent.com/29123352/80629322-4e215900-8a70-11ea-86f0-b2654f89dcaa.png)

Once done, this will resolve issue #14772.

@timabbott and @elhawk what are your thoughts on the way these messages look so far?

timabbott commented:
Thanks for working on this @Hypro999!  The formatting could use some work.

I would move the links to `[capture event](eventlink)`, `[issue](issuelink)`; there's no reason to have those URLs appear in the body as they just waste space.

And I'd put the date and time on the same line as just `time: {date} {time}`.

And we should put the traceback inside a fenced code block so that they're not mis-formatted using the "4 spaces makes a code block" feature.

Hypro999 commented:
@timabbott thanks for the review! I finished writing the integration and updating the docs and message formats - so this PR is now ready for a code review. The only thing that's left is writing the tests. I finished testing manually but held off on the automated tests for the time being in case you had any changes you wanted to suggest. I'll also update the commit message to be more informative afterwards.

The SDKs that I've tested this with are the ones for Python, Go, and Node. I'll test with other SDKs as the need arises since there are quite a lot.

Message format changes since the last review: I definitely agree that the links looked very ugly in the body. So I moved the event link accordingly and I actually got rid of the event link since that was actually an API link and not a regular web link. I also put the date and time on the same line under the key "timestamp".

> And we should put the traceback inside a fenced code block so that they're not mis-formatted using the "4 spaces makes a code block" feature.

I'm actually already doing this. I think that you might have been mis-lead by the fact that there are 3 fenced code blocks for the traceback. I did this on purpose. I wanted to separate the line that actually triggered the error/exception from the lines before and after it (the pre-context and post-context respectively). If have a better way to show this in mind, please let me know :slightly_smiling_face:.

Hypro999 commented:
Here are the screenshots that I'm using in the docs (to show the formatting updates):
![001](https://user-images.githubusercontent.com/29123352/80710460-68107980-8b0c-11ea-937b-65905ade8588.png)
![002](https://user-images.githubusercontent.com/29123352/80710464-6941a680-8b0c-11ea-8294-d5edfbe889e1.png)

timabbott commented:
> I'm actually already doing this. I think that you might have been mis-lead by the fact that there are 3 fenced code blocks for the traceback. I did this on purpose. I wanted to separate the line that actually triggered the error/exception from the lines before and after it (the pre-context and post-context respectively). If have a better way to show this in mind, please let me know slightly_smiling_face.

I see -- yeah, this is a strange way to do it.  Typically, tools handle this by adding 4 spaces of indentation before every line (to make space), and with the line having the traceback, instead `--->`.  That makes it very visibly clear what the source of the error is.

Hypro999 commented:
@timabbott, sounds like a good idea. I've done exactly that. So how does the code look? Once I get the thumbs up from you, I'll add the missing automated tests and edit the commit message.

Hypro999 commented:
Here's what it looks like now:
![001](https://user-images.githubusercontent.com/29123352/80787781-18ca5780-8ba5-11ea-9dc9-7644235bbb49.png)

timabbott commented:
Yeah, that looks great!

Hypro999 commented:
Hmm, there seems to have been a flake in the test:
```
May 02 06:13:42 Installing Yarn!
May 02 06:13:42 Downloading tarball...
May 02 06:13:42 
May 02 06:13:42 [1/2]: https://yarnpkg.com/downloads/1.22.4/yarn-v1.22.4.tar.gz --> /tmp/yarn.tar.gz.EvIQOiLcII
May 02 06:13:42   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
May 02 06:13:42                                  Dload  Upload   Total   Spent    Left  Speed
100    80  100    80    0     0   1818      0 --:--:-- --:--:-- --:--:--  1818
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
May 02 06:13:43 curl: (51) SSL: no alternative certificate subject name matches target host name 'classic.yarnpkg.com'
May 02 06:13:43 
May 02 06:13:43 [2/2]: https://yarnpkg.com/downloads/1.22.4/yarn-v1.22.4.tar.gz.asc --> /tmp/yarn.tar.gz.EvIQOiLcII.asc
100    84  100    84    0     0  10500      0 --:--:-- --:--:-- --:--:-- 10500
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
May 02 06:13:43 curl: (51) SSL: no alternative certificate subject name matches target host name 'classic.yarnpkg.com'
May 02 06:13:43 > Failed to download https://yarnpkg.com/downloads/1.22.4/yarn-v1.22.4.tar.gz.
May 02 06:13:43 
May 02 06:13:43 Error running a subcommand of ./lib/provision.py: sudo -H -- env http_proxy= https_proxy= no_proxy= scripts/lib/install-node
May 02 06:13:43 Actual error output for the subcommand is just above this.
May 02 06:13:43 
May 02 06:13:43 Traceback (most recent call last):
May 02 06:13:43   File "./lib/provision.py", line 432, in <module>
May 02 06:13:43     main(options)
May 02 06:13:43   File "./lib/provision.py", line 358, in main
May 02 06:13:43     run_as_root(proxy_env + ["scripts/lib/install-node"], sudo_args = ['-H'])
May 02 06:13:43   File "/home/circleci/zulip/scripts/lib/zulip_tools.py", line 450, in run_as_root
May 02 06:13:43     run(args, **kwargs)
May 02 06:13:43   File "/home/circleci/zulip/scripts/lib/zulip_tools.py", line 189, in run
May 02 06:13:43     subprocess.check_call(args, **kwargs)
May 02 06:13:43   File "/usr/lib/python3.6/subprocess.py", line 311, in check_call
May 02 06:13:43     raise CalledProcessError(retcode, cmd)
May 02 06:13:43 subprocess.CalledProcessError: Command '['sudo', '-H', '--', 'env', 'http_proxy=', 'https_proxy=', 'no_proxy=', 'scripts/lib/install-node']' returned non-zero exit status 1.
May 02 06:13:43 
May 02 06:13:43 Provisioning failed (exit code 1)!
May 02 06:13:43 
May 02 06:13:43 * Look at the traceback(s) above to find more about the errors.
May 02 06:13:43 * Resolve the errors or get help on chat.
May 02 06:13:43 * If you can fix this yourself, you can re-run tools/provision at any time.
May 02 06:13:43 * Logs are here: zulip/var/log/provision.log
May 02 06:13:43 
+ ret=1
+ '[' 1 = 1 ']'
+ echo '`provision`: Something went wrong with the provisioning, might be a network issue, Retrying to provision...'
+ tools/provision
May 02 06:13:43 `provision`: Something went wrong with the provisioning, might be a network issue, Retrying to provision...
May 02 06:13:43 No changes to apt dependencies, so skipping apt operations.
May 02 06:13:43 + sudo -H -- env http_proxy= https_proxy= no_proxy= scripts/lib/install-node
May 02 06:13:43 Installing Yarn!
May 02 06:13:43 Downloading tarball...
May 02 06:13:43 
May 02 06:13:43 [1/2]: https://yarnpkg.com/downloads/1.22.4/yarn-v1.22.4.tar.gz --> /tmp/yarn.tar.gz.3OSPE0SGAt
May 02 06:13:43   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
May 02 06:13:43                                  Dload  Upload   Total   Spent    Left  Speed
100    80  100    80    0     0   2222      0 --:--:-- --:--:-- --:--:--  2222
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
May 02 06:13:43 curl: (51) SSL: no alternative certificate subject name matches target host name 'classic.yarnpkg.com'
May 02 06:13:43 
May 02 06:13:43 [2/2]: https://yarnpkg.com/downloads/1.22.4/yarn-v1.22.4.tar.gz.asc --> /tmp/yarn.tar.gz.3OSPE0SGAt.asc
100    84  100    84    0     0  10500      0 --:--:-- --:--:-- --:--:-- 10500
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
May 02 06:13:43 curl: (51) SSL: no alternative certificate subject name matches target host name 'classic.yarnpkg.com'
May 02 06:13:43 > Failed to download https://yarnpkg.com/downloads/1.22.4/yarn-v1.22.4.tar.gz.
May 02 06:13:43 
May 02 06:13:43 Error running a subcommand of ./lib/provision.py: sudo -H -- env http_proxy= https_proxy= no_proxy= scripts/lib/install-node
May 02 06:13:43 Actual error output for the subcommand is just above this.
May 02 06:13:43 
May 02 06:13:43 Traceback (most recent call last):
May 02 06:13:43   File "./lib/provision.py", line 432, in <module>
May 02 06:13:43     main(options)
May 02 06:13:43   File "./lib/provision.py", line 358, in main
May 02 06:13:43     run_as_root(proxy_env + ["scripts/lib/install-node"], sudo_args = ['-H'])
May 02 06:13:43   File "/home/circleci/zulip/scripts/lib/zulip_tools.py", line 450, in run_as_root
May 02 06:13:43     run(args, **kwargs)
May 02 06:13:43   File "/home/circleci/zulip/scripts/lib/zulip_tools.py", line 189, in run
May 02 06:13:43     subprocess.check_call(args, **kwargs)
May 02 06:13:43   File "/usr/lib/python3.6/subprocess.py", line 311, in check_call
May 02 06:13:43     raise CalledProcessError(retcode, cmd)
May 02 06:13:43 subprocess.CalledProcessError: Command '['sudo', '-H', '--', 'env', 'http_proxy=', 'https_proxy=', 'no_proxy=', 'scripts/lib/install-node']' returned non-zero exit status 1.
May 02 06:13:43 
May 02 06:13:43 Provisioning failed (exit code 1)!
May 02 06:13:43 
May 02 06:13:43 * Look at the traceback(s) above to find more about the errors.
May 02 06:13:43 * Resolve the errors or get help on chat.
May 02 06:13:43 * If you can fix this yourself, you can re-run tools/provision at any time.
May 02 06:13:43 * Logs are here: zulip/var/log/provision.log
May 02 06:13:43 

Exited with code exit status 1
```

Force pushing to re-trigger the build...

Hypro999 commented:
@timabbott, this PR is ready for a final round of review, looks like our yarn mirror's certificate expired though so the tests are failing during provision.

In the most recent push, I've added the automated tests, made some really minor tweaks in the code (to get rid of trailing whitespace in the tracebacks) and updated the commit message.

Hypro999 commented:
Ok, now all of the tests pass.

timabbott commented:
Looks great, merged, thanks @Hypro999!

Hypro999 commented:
Great! Thanks for the review!
